pkg/haproxy/haproxy.go

```diff
@@ -128,42 +145,58 @@ func (h *HAProxySetManager) StopAll() {
 	h.ctx, h.cxl = context.WithCancel(h.parentCtx)
 }
 
-func (h *HAProxySetManager) StopOne(listenAddr string) {
+// StopOne key generated by GetRemovals which iterates over h.sources
+// call stop with prebuilt key
+func (h *HAProxySetManager) StopOne(listenAddrWithPort string) {
 	h.Lock()
 	defer h.Unlock()
-	h.logger.Debugf("StopOne called for %v", listenAddr)
+	h.logger.Debugf("StopOne called for %v", listenAddrWithPort)
 
-	if cxl, ok := h.cancelFuncs[listenAddr]; !ok {
+	if cxl, ok := h.cancelFuncs[listenAddrWithPort]; !ok {
 		return
 	} else {
 		cxl()
+		delete(h.cancelFuncs, listenAddrWithPort)
+		delete(h.sources, listenAddrWithPort)
 	}
 }
 
+// Configure creates an haproxy config from a given v6 backend and set of pods.
+// It makes a config for each v6-addr:port combination the reason for this is
+// that when a pod supporting a VIP goes away, the config MUST be rewritten so
+// we don't route traffic to dead pod IPs. When this happens, HAProxy must reload
+// the config which causes a (very brief) downtime for a server. A 1:1 map from
+// backend to pod minimizes the detrimental effect of pod churn from a different
+// service reloading a config for a service that isn't relevant to it
 func (h *HAProxySetManager) Configure(config VIPConfig) error {
 	listenAddr := config.Addr6
-	serviceAddrs := config.ServiceAddrs
-	ports := config.ListenPorts
+	podIPs := config.PodIPs
+	targetPort := config.TargetPort
+	servicePort := config.ServicePort
 
-	h.logger.Debugf("configuring s=%v d=%v p=%v", listenAddr, serviceAddrs, ports)
+	h.logger.Debugf("configuring s=%v d=%v tPort=%v sPort=%v", listenAddr, podIPs, targetPort, servicePort)
 	h.Lock()
 	defer h.Unlock()
 
 	// create the instance if it doesn't exist
-	if _, found := h.sources[listenAddr]; !found {
+	if _, found := h.sources[h.createInstanceKey(listenAddr, servicePort)]; !found {
 		c2, cxl := context.WithCancel(h.ctx)
-		instance, err := NewHAProxy(c2, h.binary, h.configDir, listenAddr, serviceAddrs, ports, h.errChan, h.logger)
+		instance, err := NewHAProxy(c2, h.binary, h.configDir, listenAddr, podIPs, targetPort, servicePort, h.errChan, h.logger)
 		if err != nil {
 			h.logger.Errorf("error creating new haproxy. canceling context. %v", err)
 			cxl()
 			return err
 		}
-		h.sources[listenAddr] = instance
-		h.cancelFuncs[listenAddr] = cxl
+		h.sources[h.createInstanceKey(listenAddr, servicePort)] = instance
+		h.cancelFuncs[h.createInstanceKey(listenAddr, servicePort)] = cxl
 	}
 
 	// then configure it
-	return h.sources[listenAddr].Reload(ports)
+	return h.sources[h.createInstanceKey(listenAddr, servicePort)].Reload(podIPs, targetPort, servicePort)
```

Comment:
can you get a value for this `h.createInstanceKey(listenAddr, servicePort)` early in the function and reuse it?

Reply:
The Reload() fn has some conditional checks that might result in it doing nothing now that the structure of the file has changed.

Reply:
1. done 2. Not sure, will test cases to flesh this out